File: opacus-sdk/src/crypto/keys.ts
/**
 * Opacus Key Manager
 * Handles Ed25519 signing keys and X25519 encryption keys
 */

import * as ed from '@noble/ed25519';
import { x25519 } from '@noble/curves/ed25519';
import { sha256 } from '@noble/hashes/sha256';
import { randomBytes } from '@noble/hashes/utils';
import { AgentIdentity } from '../types';

export class KeyManager {
  /**
   * Generate Ed25519 key pair for signatures
   */
  static async generateEd25519(): Promise<{pub: Uint8Array, priv: Uint8Array}> {
    const priv = ed.utils.randomPrivateKey();
    const pub = await ed.getPublicKeyAsync(priv);
    return { priv, pub };
  }

  /**
   * Generate X25519 key pair for ECDH
   */
  static generateX25519(): {pub: Uint8Array, priv: Uint8Array} {
    const priv = randomBytes(32);
    const pub = x25519.getPublicKey(priv);
    return { priv, pub };
  }

  /**
   * Generate full agent identity with both key pairs
   */
  static async generateFullIdentity(chainId = 16602): Promise<AgentIdentity> {
    const edKeys = await this.generateEd25519();
    const xKeys = this.generateX25519();
    const id = this.toHex(sha256(edKeys.pub).slice(0, 20));
    const address = '0x' + this.toHex(sha256(edKeys.pub).slice(0, 20));
    
    return {
      id,
      edPub: edKeys.pub,
      edPriv: edKeys.priv,
      xPub: xKeys.pub,
      xPriv: xKeys.priv,
      address,
      chainId
    };
  }

  /**
   * Convert bytes to hex string
   */
  static toHex(bytes: Uint8Array): string {
    return Array.from(bytes).map(b => b.toString(16).padStart(2, '0')).join('');
  }

  /**
   * Convert hex string to bytes
   */
  static fromHex(hex: string): Uint8Array {
    const clean = hex.startsWith('0x') ? hex.slice(2) : hex;
    const bytes = new Uint8Array(clean.length / 2);
    for (let i = 0; i < bytes.length; i++) {
      bytes[i] = parseInt(clean.substr(i * 2, 2), 16);
    }
    return bytes;
  }

  /**
   * Export identity to JSON
   */
  static exportIdentity(identity: AgentIdentity): string {
    return JSON.stringify({
      id: identity.id,
      edPub: this.toHex(identity.edPub),
      edPriv: this.toHex(identity.edPriv),
      xPub: this.toHex(identity.xPub),
      xPriv: this.toHex(identity.xPriv),
      address: identity.address,
      chainId: identity.chainId
    });
  }

  /**
   * Import identity from JSON
   */
  static importIdentity(json: string): AgentIdentity {
    const data = JSON.parse(json);
    return {
      id: data.id,
      edPub: this.fromHex(data.edPub),
      edPriv: this.fromHex(data.edPriv),
      xPub: this.fromHex(data.xPub),
      xPriv: this.fromHex(data.xPriv),
      address: data.address,
      chainId: data.chainId
    };
  }
}

File: opacus-rust/src/crypto/keys.rs
//! Key generation and management

use ed25519_dalek::{SigningKey, VerifyingKey};
use x25519_dalek::{StaticSecret, PublicKey as X25519Public};
use sha2::{Sha256, Digest};
use rand::rngs::OsRng;
use crate::types::AgentIdentity;

/// Key manager for Ed25519 and X25519 operations
pub struct KeyManager;

impl KeyManager {
    /// Generate Ed25519 key pair for signing
    pub fn generate_ed25519() -> (SigningKey, VerifyingKey) {
        let signing_key = SigningKey::generate(&mut OsRng);
        let verifying_key = signing_key.verifying_key();
        (signing_key, verifying_key)
    }
    
    /// Generate X25519 key pair for encryption
    pub fn generate_x25519() -> (StaticSecret, X25519Public) {
        let secret = StaticSecret::random_from_rng(OsRng);
        let public = X25519Public::from(&secret);
        (secret, public)
    }
    
    /// Generate full agent identity with dual keys
    /// 
    /// # Arguments
    /// * `chain_id` - Blockchain chain ID
    /// 
    /// # Returns
    /// Complete `AgentIdentity` with Ed25519 and X25519 keys
    pub fn generate_identity(chain_id: u64) -> AgentIdentity {
        let (ed_signing, ed_verifying) = Self::generate_ed25519();
        let (x_secret, x_public) = Self::generate_x25519();
        
        // Generate ID from public key hash
        let mut hasher = Sha256::new();
        hasher.update(ed_verifying.as_bytes());
        let hash = hasher.finalize();
        let id = hex::encode(&hash[..20]);
        let address = format!("0x{}", hex::encode(&hash[..20]));
        
        AgentIdentity {
            id,
            ed_pub: *ed_verifying.as_bytes(),
            ed_priv: ed_signing.to_bytes(),
            x_pub: x_public.to_bytes(),
            x_priv: x_secret.to_bytes(),
            address,
            chain_id,
        }
    }
    
    /// Convert bytes to hex string
    pub fn to_hex(bytes: &[u8]) -> String {
        hex::encode(bytes)
    }
    
    /// Parse hex string to bytes
    pub fn from_hex(s: &str) -> Result<Vec<u8>, hex::FromHexError> {
        hex::decode(s)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    
    #[test]
    fn test_generate_ed25519() {
        let (signing, verifying) = KeyManager::generate_ed25519();
        assert_eq!(signing.to_bytes().len(), 32);
        assert_eq!(verifying.as_bytes().len(), 32);
    }
    
    #[test]
    fn test_generate_x25519() {
        let (secret, public) = KeyManager::generate_x25519();
        assert_eq!(secret.to_bytes().len(), 32);
        assert_eq!(public.as_bytes().len(), 32);
    }
    
    #[test]
    fn test_generate_identity() {
        let identity = KeyManager::generate_identity(16602);
        assert_eq!(identity.chain_id, 16602);
        assert_eq!(identity.ed_pub.len(), 32);
        assert_eq!(identity.ed_priv.len(), 32);
        assert_eq!(identity.x_pub.len(), 32);
        assert_eq!(identity.x_priv.len(), 32);
        assert!(identity.address.starts_with("0x"));
    }
    
    #[test]
    fn test_hex_conversion() {
        let bytes = [1, 2, 3, 4, 5];
        let hex = KeyManager::to_hex(&bytes);
        assert_eq!(hex, "0102030405");
        let decoded = KeyManager::from_hex(&hex).unwrap();
        assert_eq!(decoded, bytes);
    }
}
